### ISSUE TITLE
Use  AppImageKit or Flatpak for Linux installation

### ISSUE THREAD

**Issue by albar965** [albar965/littlenavmap#522]
Solves library conflicts.

**Comment by MTRNord** [comment]
Hi I know this is only somewhat related but I am running steam/xp12 in flatpak and littlenavmap outside and it seems like there is no way to connect to the client due to the memory protections/sandboxing. This may also become an issue when littlenavmap itself is bundled in a flatpak. Littlenavconnect seems to also not work since the steam flatpak is missing at least `libgssapi_krb5.so.2` (and I have no idea where to get it from either).

```
➜  ~ flatpak enter com.valvesoftware.Steam /bin/bash
bash-5.1$ cd /run/media/marcel/1ac60e61-e5e2-4157-9a7a-830801a3fb0b/SteamLibrary/steamapps/common/X-Plane\ 12/Resources/plugins/Little\ Navconnect/
bash-5.1$ ./littlenavconnect 
./littlenavconnect: error while loading shared libraries: libgssapi_krb5.so.2: cannot open shared object file: No such file or directory
```

**Comment by albar965** [comment]
> Hi I know this is only somewhat related but I am running steam/xp12 in flatpak and littlenavmap outside and it seems like there is no way to connect to the client due to the memory protections/sandboxing. This may also become an issue when littlenavmap itself is bundled in a flatpak. Littlenavconnect seems to also not work since the steam flatpak is missing at least `libgssapi_krb5.so.2` (and I have no idea where to get it from either).

LNM and Xpconnect communicate using shared memory. I'd expect this to be the first thing to be blockend in a sandbox. If you can put Navconnect into the sandbox too you could use network communication. No idea if this is possible.

`libgssapi_krb5.so` is pulled in by  `libQt5Network.so.5`. This goes on my nerves too (I don't need Kerberos) but not much I can do there. No way to install the package ` libgssapi-krb5-2`?

Alex

**Comment by MTRNord** [comment]
Hi thanks knowing what package this even is from helps. I will try that. I can just compile it as part of the steam flatpak I guess to move it into it. Network based communication already works successfully for pilot as well so likely it will be working here as well :) I will report back when I tried it.

**Comment by MTRNord** [comment]
Hi after spending all day on the steam flatpak it seems like this is indeed possible to do. Only issue I am left with is the scenery db not working with xplane in network mode. But Tracking is the important part which works now. Thanks for the help.

**Comment by albar965** [comment]
You can copy the X-Plane files out of the flatpak instance out of the sandbox to the computer and then load the database outside.

Create a folder `$HOME/.x-plane` and put a file `x-plane_install_12.txt` into it. Theis file should contain only one line with the absolute path to the fake/dummy installation.

**Comment by MTRNord** [comment]
Turns out I had that file but it pointed to the XP12 demo folder which is gone instead of the steam one. Thanks again

**Comment by MTRNord** [comment]
If anyone ever comes by here in the future I documented the changes to get this working here: https://notes.midnightthoughts.space/digitalgarden/blog/using-x-plane12-on-linux-with-flatpak/ Hope that helps someone else :)

**Comment by MTRNord** [comment]
Hi, I wanted to start building a Flatpak for this while I am already deep into flatpak issues around flightsim tools :)

I was wondering if a) that is ok to do (will post a branch when I am far enough here) b) I was wondering about the license. Is it `GPL-3.0-only` or `GPL-3.0-or-later`? (I need to use a spdx license code for flatpak metadata and gpl makes a difference with that :) )

**Comment by albar965** [comment]
> a) that is ok to do (will post a branch when I am far enough here)

Perfectly fine with me if I can integrate it into the build process here. The build is done on an Ubuntu 22.04 VM, while I use Kubuntu 23.04 for development. 

> b) I was wondering about the license. Is it `GPL-3.0-only` or `GPL-3.0-or-later`?

GPL-3.0-or-later. Do you see any conflicts with the licenses?

Alex

**Comment by MTRNord** [comment]
> Perfectly fine with me if I can integrate it into the build process here.

I see so far no changes needed. Arguably the required [appstream/metainfo](https://www.freedesktop.org/software/appstream/docs/) and desktop files could be in here, but that's not strictly necessary, and they can just be within the flathub repository.

> The build is done on an Ubuntu 22.04 VM, while I use Kubuntu 23.04 for development.

The build happens within the flatpak. Due to dependencies it's probably best to use the kde sdk which I believe currently runs 22.08 with QT5.15. Newer versions only ship with QT6, which from what I had seen isn't yet supported by little navmap.

> GPL-3.0-or-later. Do you see any conflicts with the licenses?

In terms of flathub/flatpak GPL is perfectly fine :)

**Comment by MTRNord** [comment]
Hi,

The first version (not building littlenavconnect yet) is at https://github.com/MTRNord/flathub/tree/littlenavmap

I had some trouble to have the correct paths hence there is https://github.com/MTRNord/flathub/blob/littlenavmap/littlenavmap-use-flatpak-paths.patch and https://github.com/MTRNord/flathub/blob/littlenavmap/littlexpconnect-correct-deploy-dir.patch and I am not sure if there is a nicer way to do this via the upstream files.

The README points out some things like how to get the plugin from the flatpak as well as the changed config path which is relevant for navigraph. I also noticed that it seems to default to English folders in the Documents instead of localized ones like the version you provide does.

Additionally, there is one benefit for wayland users: It seems like the application now runs natively with wayland instead of using xwayland which is a nice touch.

Feel free to take a look, ask questions or recommend changes :) Building should be as easy as running `./build.sh` and then running it using the menu entry it adds. Flatpak and flatpak-builder obviously are required to build it :) (When published, there is a CI doing this on GitHub changes instead. So no need for manual building unless you want to test changes locally)

One other bug I found that is yet to fix is that for whatever reason, the version of littlexpconnect I built results in `LittleXpconnect void SharedMemoryWriter::writeData(const QByteArray&, bool) Cannot lock "" native "/dev/shm/LittleXpconnect"`. Not sure if I did something wrong, QT 5.15 did a change related to this, or if this is a flatpak issue, maybe.

**Comment by MTRNord** [comment]
Ok the lock error likely is a build fail on my side since it also reports some missing symbols on loading. I will try to fix that :)

**Comment by MTRNord** [comment]
Argh I had 2 plugins installed. One of them using some experimental patches... Obviously that fails D:

**Comment by albar965** [comment]
> I also noticed that it seems to default to English folders in the Documents instead of localized ones like the version you provide does.

This depends on found language for LNM (`LANG` or `LANGUAGE` not sure which one).

> One other bug I found that is yet to fix is that for whatever reason, the version of littlexpconnect I built results in missing symbol errors. Likely the steam flatpak doesnt have some of the required things.

Xpconnect is very special. I have to build it using a statically linked minimal version of Qt. See[
`littlenavmap/build/linux/build_qt_static.sh`.](https://github.com/albar965/littlenavmap/blob/release/2.8/build/linux/build_qt_static.sh) and https://github.com/albar965/littlenavmap/blob/af004428a880fcf16a6294a1a6f66767fdc4056b/build/linux/build_release.sh#L81-L110

Not sure about the shared memory. Does it work now?

Alex

**Comment by albar965** [comment]
Will have a look at the flatpak build but gimme a few days.

**Comment by MTRNord** [comment]
> > I also noticed that it seems to default to English folders in the Documents instead of localized ones like the version you provide does.
> 
> This depends on found language for LNM (`LANG` or `LANGUAGE` not sure which one).

I guess flatpak doesn't expose that from the host to the sandbox. I will see if there is a way to expose, it :) Would be nice to stay consistent.

> 
> > One other bug I found that is yet to fix is that for whatever reason, the version of littlexpconnect I built results in missing symbol errors. Likely the steam flatpak doesnt have some of the required things.
> 
> Xpconnect is very special. I have to build it using a statically linked minimal version of Qt. See[ `littlenavmap/build/linux/build_qt_static.sh`.](https://github.com/albar965/littlenavmap/blob/release/2.8/build/linux/build_qt_static.sh) and
> 
> https://github.com/albar965/littlenavmap/blob/af004428a880fcf16a6294a1a6f66767fdc4056b/build/linux/build_release.sh#L81-L110
> 
> Not sure about the shared memory. Does it work now?

I am actually struggling to get xpconnect working :D The plugin I am getting currently fails with missing qt symbols and so on even with a static qt build. I am surely not doing the right thing yet :) It also wants the specific libicu so version of the flatpak it was built in. So its not portable like the version you provide. I saw the mention about static qt in your release file but didnt see the way you build the qt. Maybe there is the magic that I am missing. I will give it a try :)

> 
> Alex

**Comment by albar965** [comment]
You have to build the atools with the `ATOOLS_NO...=true` environment variables to get rid of GUI stuff. Maybe this is the reason.

**Comment by MTRNord** [comment]
> You have to build the atools with the `ATOOLS_NO...=true` environment variables to get rid of GUI stuff. Maybe this is the reason.

I think I was looking at the wrong version of your release file and only have NO_GRIB and NO_FS set to true :) Thats probably the reason then.

**Comment by MTRNord** [comment]
> Not sure about the shared memory. Does it work now?

Hm it seems to still use the /tmp folder instead of /dev/shm.

```
[📦 com.valvesoftware.Steam ~]$ ls -la /tmp
insgesamt 6156
drwx------.  7 marcel marcel     600 16. Jul 15:29 .
drwxr-xr-x. 11 marcel marcel     340 16. Jul 15:28 ..
-rw-r-----.  1 marcel marcel       0 16. Jul 15:26 qipc_sharedmemory_LittleXpconnect95a51c8248d6bdb605dc1935e50241bbbbc64b65
-rw-r-----.  1 marcel marcel       0 16. Jul 15:26 qipc_systemsem_LittleXpconnect95a51c8248d6bdb605dc1935e50241bbbbc64b65
```

But at least the plugin now works. Thats already a step in a good direction from where I was at before :)

**Comment by MTRNord** [comment]
Also it seems like the static qt build isnt needed somehow for flatpak. It worked just fine when using the qt version which the flatpak sdk ships with which isnt static linked.

**Comment by MTRNord** [comment]
Ok turns out that I was going down the wrong road. /dev/shm will never be used by qt. And since the sharedmemory native path is not going to be predicatble I dont think we can expose just the 2 files of it. One fix that works but is ugly is to expose /tmp to the host. which then shares the sharedmemory of course.  I will go and ask the flatpak people if someone dealt with this before

**Comment by albar965** [comment]
Keep in mind that Xpconnect is not supposed to run in the flatpak sandbox. It's gotta be copied into X-Plane plugins and it has to be built statically to avoid issues with other plugins on Linux.

Running Navconnect and Xpconnect outside the box with the sim is also no option. You'd still need a not flatpaked Navconnect. I'd rather expose /tmp

**Comment by MTRNord** [comment]
> Keep in mind that Xpconnect is not supposed to run in the flatpak sandbox. It's gotta be copied into X-Plane plugins and it has to be built statically to avoid issues with other plugins on Linux.

yes my current test setup is X-Plane and xpconnect running in the steam flatpak and littlenavmap in the one I am writing.



> Running Navconnect and Xpconnect outside the box with the sim is also no option. You'd still need a not flatpaked Navconnect. I'd rather expose /tmp

Yep thats not nice either.

My only worry with exposing /tmp is that it may not be accepted by the flathub people. Since it kind of defeats the idea of sandboxing the application. Ideally I would just expose the shared memory files. But it seems like flatpak can only expose folders but not files alone. So for now exposing /tmp it is until I come up with a better solution or later in a review from flathub people someone has a good idea to do this instead. Either way in terms of the steam flatpak this needs an extra step to work anyway to expose /tmp. But that will be documented in the readme.

**Comment by MTRNord** [comment]
> > Keep in mind that Xpconnect is not supposed to run in the flatpak sandbox. It's gotta be copied into X-Plane plugins and it has to be built statically to avoid issues with other plugins on Linux.
> 
> yes my current test setup is X-Plane and xpconnect running in the steam flatpak and littlenavmap in the one I am writing.

Actually I guess I should also test it later today with a stock steam flatpak instead of my modified one. I just realised that I may have more QT libs available than in the upstream version which made it work.

**Comment by jornfranke** [comment]
Is there an update or some ideas what blocks it?

Flatpak is very useful for people that use immutable Linux Desktops (e.g. Aeon, KalpaDesktop, Bazzite etc.

**Comment by albar965** [comment]
> Is there an update or some ideas what blocks it?
> 
> Flatpak is very useful for people that use immutable Linux Desktops (e.g. Aeon, KalpaDesktop, Bazzite etc.

Time. 🤷‍♂️

**Comment by jornfranke** [comment]
Ok, I understand. Is thee any way one can help?

**Comment by albar965** [comment]
> Ok, I understand. Is thee any way one can help?

Yes. Thank you for your offer. I'm not sure if @MTRNord is still active here. Seriously appreciate any work done by @MTRNord so far.
I pushed this issue to low priority since I have too much to do in RL and LNM too (Qt 6 porting fun, user support and more).

I remember various issues (also above) with included plugins and programs. 

Some points that can cause issues:
- LNM has an option to install the included Xpconnect plugin into X-Plane from LNM menu `Tools`. Not sure if this needs work on the Flatpak package.
- LNM communicates with the Xpconnect plugin using shared memory. This probably needs a hole punched into the sandbox. Looks like it's already done (see above).
- For remote computers access to the local network is needed for LNM.
- The Navconnect program is included in the program and needs open network connection for listening as well as shared memory access to Xpconnect. No idea how two programs can fit into one package.

This all needs testing and probably modifications. I have only a rough idea about Flatpak. So any help is welcome.

There is a script that integrates into my build system which creates the Debian package:
[create_deb.sh](https://github.com/albar965/littlenavmap/blob/release/3.0/build/linux/create_deb.sh)

A similar script for Flatpak would allow me to add this to one of the next builds.

Note that current work is done on branch `release/3.0` or `release/4.0` for atools.

Alex

**Comment by MTRNord** [comment]
While I am still reading here, I did move on from this. :) Not sure how much use the patches in https://github.com/MTRNord/flathub/tree/littlenavmap/patch still have but that repo is where I left it of. It was usable at the time and I basically used it all the time. I just never really go beyond that or update it since.

**Comment by albar965** [comment]
> While I am still reading here, I did move on from this. :) Not sure how much use the patches in https://github.com/MTRNord/flathub/tree/littlenavmap/patch still have but that repo is where I left it of. It was usable at the time and I basically used it all the time. I just never really go beyond that or update it since.

Ok. I'll have a look at this but no idea when.

**Comment by lostiniceland** [comment]
Would be nice to find this app on flathub.

**Comment by vibhavp** [comment]
I've been using this manifest to build the latest release: https://github.com/vibhavp/littlenavmap-flatpak. The build includes Xpconnect and has the appropriate perms to install this to an X-Plane installation (regardless of whether X-Plane/Steam is sandboxed or not), and talk to the Xpconnect plugin with `--filesystem=tmp`.It also includes the default Navigraph database bundled with the official release. I'd be more than happy to donate these files to the project so that the Flatpak build can be maintained in the same repository.
